Auth with API Key

johnswanson opened the pull request:
In terms of the actual application code, things are pretty straightforward: if an API key is passed, find the associated user, check the hash against the passed API key, and attach the user if everything went well.

I renamed the existing functions related to API keys, e.g. `mw.auth/wrap-api-key`. We should be careful about overloading the term "API Key", both internally and externally.

(The "API Key" for those existing functions is a static key defined by an env var that allows you to require a particular header *in addition to regular authentication* on some endpoints.)

It's probably too painful to change the actual setting name, but we can change what we call it in the code.

Fixes #36570

johnswanson committed:
Auth with API Key

In terms of the actual application code, things are pretty straightforward: if an API key is passed, find the
associated user, check the hash against the passed API key, and attach the user if everything went well.

johnswanson committed:
rename existing API key functions to make distinction more clear

We have existing functions related to the `MB_API_KEY` setting. This allows users to set a static API key that can
provide an additional layer of authentication to some routes.

I'm keeping the setting name unchanged to avoid issues with backwards incompatibility.

johnswanson committed:
Add a high level test for API key authentication

While we have existing tests at the middleware level to verify that API key authentication works, I wanted to add a
higher-level test that we could create a new API key and then authenticate using it.

noahmoss commented on test/metabase/api/api_key_test.clj:
```diff
@@ -102,3 +103,14 @@
                             :created_by   (mt/user->id :crowberto)}]
         (is (= 2 (mt/user-http-request :crowberto :get 200 "api-key/count")))))))
+
+(deftest api-keys-work-e2e
```
Could you add a test that all of the session-level dynamic vars (like `api/is-superuser?`, etc) are bound correctly for API users?

noahmoss commented on src/metabase/server/middleware/session.clj:
```diff
@@ -305,17 +324,41 @@
+  (u.password/verify-password api-key-that-should-never-match "" hash-that-should-never-match))
+
+(defn- matching-api-key? [{:keys [api-key] :as _user-data} passed-api-key]
+  ;; if we get an API key, check the hash against the passed value. If not, don't reveal info via a timing attack - do
+  ;; a useless hash, *then* return `false`.
```
awesome

noahmoss commented on test/metabase/server/middleware/session_test.clj:
```diff
@@ -243,6 +243,25 @@
 
+(deftest current-user-info-for-api-key-test
+  (try
+    (t2/insert! :model/ApiKey {:id 999
+                               :name "An API Key"
+                               :user_id (mt/user->id :lucky)
+                               :created_by (mt/user->id :lucky)
+                               :unhashed_key "mb_foobar"})
```
Can you use `t2.with-temp` here instead?

johnswanson commented on src/metabase/server/middleware/session.clj:
```diff
@@ -290,6 +292,23 @@
+   (t2.pipeline/compile*
+    {:select    [[:api_key.user_id :metabase-user-id]
+                 [:api_key.key :api-key]
+                 [:user.is_superuser :is-superuser?]
+                 [:user.locale :user-locale]]
```
Good point - this actually raises a question: can API key users become group managers, and if so, how? My thinking was that API key users will be excluded from the users you get back from normal user API endpoints, so you wouldn't be able to go through the normal flow for making them a group manager. But, that's maybe orthogonal to the question of "if a API key user _is_ a group manager, do we respect that?" - which I think the answer is yes, we should. I'll change this.

(And then maybe at some point down the road we'll add the UI/API necessary to actually make an API key into a group manager.)

johnswanson committed:
CR response

- update the comment in `metabase.server.middleware.session`:

  - talk about API key auth

  - remove outdated paragraph about `metabase.SESSION_ID` (this seems to
  have been removed)

  - turn it into a namespace docstring instead of a comment

- select `is-group-manager?` for API keys

- add a test that session-level dynamic vars are bound correctly for API
keys

- switch to `t2.with-temp` in tests

johnswanson committed:
more CR response!

johnswanson commented on src/metabase/server/middleware/session.clj:
```diff
@@ -305,17 +333,43 @@
+(defn- current-user-info-for-api-key
+  "Return User ID and superuser status for an API Key with `api-key-id"
+  [api-key]
+  (when api-key
```
Good call, done!

johnswanson commented on src/metabase/server/middleware/session.clj:
```diff
@@ -305,17 +333,43 @@
   (merge
    request
    (current-user-info-for-session metabase-session-id anti-csrf-token)
+   (current-user-info-for-api-key x-api-key)
```
Good idea, done - I used an `or` so we'll merge with the first non-nil user-info, if any (with the session taking precedence over the API key).
